tls/s2n_post_handshake.c

```diff
@@ -13,66 +13,158 @@
  * permissions and limitations under the License.
  */
 
+#include <sys/param.h>
+
 #include "error/s2n_errno.h"
 
 #include "tls/s2n_connection.h"
 #include "tls/s2n_key_update.h"
 #include "tls/s2n_tls.h"
 #include "utils/s2n_safety.h"
 
-int s2n_post_handshake_recv(struct s2n_connection *conn) 
+static S2N_RESULT s2n_post_handshake_process(struct s2n_connection *conn, struct s2n_stuffer *in, uint8_t message_type)
 {
-    POSIX_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(conn);
 
-    uint8_t post_handshake_id;
-    uint32_t message_length;
+    switch (message_type) {
+        case TLS_KEY_UPDATE:
+            RESULT_GUARD_POSIX(s2n_key_update_recv(conn, in));
+            break;
+        case TLS_SERVER_NEW_SESSION_TICKET:
+            RESULT_GUARD(s2n_tls13_server_nst_recv(conn, in));
+            break;
+        case TLS_HELLO_REQUEST:
+            RESULT_GUARD(s2n_client_hello_request_recv(conn));
+            break;
+        case TLS_CERT_REQ:
+            /*
+             * s2n-tls does not support post-handshake authentication.
+             *
+             *= https://tools.ietf.org/rfc/rfc8446#section-4.6.2
+             *# A client that receives a CertificateRequest message without having
+             *# sent the "post_handshake_auth" extension MUST send an
+             *# "unexpected_message" fatal alert.
+             */
+            RESULT_BAIL(S2N_ERR_BAD_MESSAGE);
+        default:
+            /* All other messages are unexpected */
+            RESULT_BAIL(S2N_ERR_BAD_MESSAGE);
```

Comment:
It simplifies the logic, since we don't need to ignore messages without allocating memory for them, and seems to be the standard behavior.

The RFC isn't explicit about how unexpected post-handshake messages should be treated, but it does generally say for handshake messages:
> Protocol messages MUST be sent in the order defined in [Section 4.4.1](https://www.rfc-editor.org/rfc/rfc8446#section-4.4.1)
   and shown in the diagrams in [Section 2](https://www.rfc-editor.org/rfc/rfc8446#section-2).  A peer which receives a
   handshake message in an unexpected order MUST abort the handshake
   with an "unexpected_message" alert.

Any future post-handshake messages will probably require an extension to indicate support, like with post-handshake auth.

I've verified that Openssl responds to unknown messages with a fatal "unexpected_message" alert.